src/mr_allocator.rs

```diff
@@ -1,60 +1,603 @@
-            &self.pd,
-            addr,
-            layout.size(),
-            access,
-        )?);
-        Ok(LocalMr::new(raw_mr.addr(), raw_mr.length(), raw_mr))
+        self.alloc_from_je(layout).map_or_else(||{
+            Err(io::Error::new(io::ErrorKind::OutOfMemory, "insufficient contiguous memory was available to service the allocation request"))
+        }, |addr|{
+            let raw_mr = self.lookup_raw_mr(addr as usize);
+            Ok(LocalMr::new(addr as usize, layout.size(), raw_mr))
+        })
+    }
+
+    /// Alloc memory for RDMA operations from jemalloc
+    #[allow(clippy::as_conversions)]
+    fn alloc_from_je(&self, layout: &Layout) -> Option<*mut u8> {
+        let addr = unsafe {
+            tikv_jemalloc_sys::mallocx(
+                layout.size(),
+                (MALLOCX_ALIGN(layout.align()) | MALLOCX_ARENA(self.arena_ind.cast())).cast(),
+            )
+        };
+        if addr.is_null() {
+            None
+        } else {
+            Some(addr.cast::<u8>())
+        }
+    }
+
+    /// Look up `raw_mr` info by addr
+    #[allow(clippy::unreachable)]
+    fn lookup_raw_mr(&self, addr: usize) -> Arc<RawMemoryRegion> {
+        EXTENT_TOKEN_MAP.map_read(|map| {
+            map.range(..addr.overflow_add(1)).next_back().map_or_else(
+                || {
+                    unreachable!("can not find raw mr by addr {}", addr);
+                },
+                |(_, item)| {
+                    debug!("LOOK addr {}, item {:?}", addr, item);
+                    assert_eq!(self.arena_ind, item.arena_ind);
+                    assert!(addr >= item.addr && addr < item.addr.overflow_add(item.len));
+                    Arc::<RawMemoryRegion>::clone(&item.raw_mr)
+                },
+            )
+        })
+    }
+}
+
+/// Parking lot mutex wrapper.
+/// Provides functional expression methods.
+pub(crate) struct ParkingLotMutex<T> {
+    /// Parking_lot mutex
+    inner: Mutex<T>,
+}
+
+impl<T> ParkingLotMutex<T> {
+    /// New a `ParkingLotMutex`
+    fn new(v: T) -> Self {
+        Self {
+            inner: Mutex::new(v),
+        }
+    }
+
+    /// Use `func` to read the value in the mutex
+    fn map_read<F, R>(&self, func: F) -> R
+    where
+        F: FnOnce(&MutexGuard<'_, T>) -> R,
+    {
+        let mut guard = self.inner.try_lock();
+        while guard.is_none() {
+            guard = self.inner.try_lock();
+        }
+        // guard is checked
+        #[allow(clippy::unwrap_used)]
+        func(&guard.unwrap())
+    }
+
+    /// Use `func` to write the value in the mutex
+    fn map_write<F, R>(&self, func: F) -> R
+    where
+        F: FnOnce(&mut MutexGuard<'_, T>) -> R,
+    {
+        let mut guard = self.inner.try_lock();
+        while guard.is_none() {
+            guard = self.inner.try_lock();
+        }
+        // guard is checked
+        #[allow(clippy::unwrap_used)]
+        func(&mut guard.unwrap())
+    }
```

Comment:
Why use spinning here?

Reply:
Parkign_lot mutex will block worker threads waiting for the lock to become available.

Reply:
> Parkign_lot mutex will block worker threads waiting for the lock to become available.

`parking_lot` has implemented [adaptive spinning for micro-contention](https://docs.rs/parking_lot/latest/parking_lot/type.Mutex.html#differences-from-the-standard-library-mutex). There is no need to write extra spinning.

Reply:
> > Parkign_lot mutex will block worker threads waiting for the lock to become available.
> 
> `parking_lot` has implemented [adaptive spinning for micro-contention](https://docs.rs/parking_lot/latest/parking_lot/type.Mutex.html#differences-from-the-standard-library-mutex). There is no need to write extra spinning.

You're right! I read some code of `parking_lot::Mutex` and found that its `SpinWait` implementation is more efficient.
Thank you for prompt!